[Flaky test] Coverage test failing intermittently due to panic

*Description*:
Coverage test failing due to panic in goroutine after `TestConfigLoaderStandaloneExtensionServerAndCustomResource` has completed. Retry succeeds.

*Repro steps*:
https://github.com/envoyproxy/gateway/actions/runs/24496848059/job/71593709753

*Environment*:
CI

*Logs*:
```
1.7763234386184103e+09	info	config-loader	loader/configloader.go:119	running hook
panic: Log in goroutine after TestConfigLoaderStandaloneExtensionServerAndCustomResource has completed: config changed 3 times
	

goroutine 25 [running]:
testing.(*common).log(0xc00030a248, {0xc0000461b0, 0x16})
	/opt/hostedtoolcache/go/1.26.2/x64/src/testing/testing.go:1039 +0x1df
testing.(*common).Logf(0xc00030a248, {0x1a0730f, 0x17}, {0xc000318e88, 0x1, 0x1})
	/opt/hostedtoolcache/go/1.26.2/x64/src/testing/testing.go:1200 +0x8f
github.com/envoyproxy/gateway/internal/envoygateway/config/loader.TestConfigLoaderStandaloneExtensionServerAndCustomResource.func3({0x1aa72d8?, 0xc00017c190?}, 0xc0001a4090)
	/home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader_test.go:93 +0xc5
github.com/envoyproxy/gateway/internal/envoygateway/config/loader.(*Loader).runHook.func1({0x1aa72d8, 0xc00017c190})
	/home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader.go:129 +0x13e
created by github.com/envoyproxy/gateway/internal/envoygateway/config/loader.(*Loader).runHook in goroutine 44
	/home/runner/work/gateway/gateway/internal/envoygateway/config/loader/configloader.go:124 +0x325
FAIL	github.com/envoyproxy/gateway/internal/envoygateway/config/loader	0.110s
```

--- a/internal/envoygateway/config/loader/configloader_test.go
+++ b/internal/envoygateway/config/loader/configloader_test.go
@@ -11,6 +11,7 @@ import (
 	"os"
 	"sync/atomic"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/require"
 
@@ -31,104 +32,71 @@ var (
 	minimalConfig string
 )
 
-func TestConfigLoader(t *testing.T) {
-	tmpDir, err := os.MkdirTemp("", "envoy-gateway-configloader-test")
-	require.NoError(t, err)
-	defer func(path string) {
-		_ = os.RemoveAll(path)
-	}(tmpDir)
+const (
+	reloadTimeout = 30 * time.Second
+	reloadTick    = 10 * time.Millisecond
+)
 
+func TestConfigLoader(t *testing.T) {
+	tmpDir := t.TempDir()
 	cfgPath := tmpDir + "/config.yaml"
 	require.NoError(t, os.WriteFile(cfgPath, []byte(defaultConfig), 0o600))
 	s, err := config.New(os.Stdout, os.Stderr)
 	require.NoError(t, err)
 
-	ctx, cancel := context.WithCancel(context.Background())
-	defer func() {
-		cancel()
-	}()
-
-	changed := 0
-	loader := New(cfgPath, s, func(hookCtx context.Context, _ *config.Server) error {
-		changed++
-		// Only log if the hook context is still active to avoid panic
-		select {
-		case <-hookCtx.Done():
-			return nil
-		default:
-			t.Logf("config changed %d times", changed)
-		}
-		if changed > 1 {
-			cancel()
-		}
+	var reloads atomic.Int32
+	loader := New(cfgPath, s, func(context.Context, *config.Server) error {
+		reloads.Add(1)
 		return nil
 	})
 
-	require.NoError(t, loader.Start(ctx, os.Stdout))
-	go func() {
-		_ = os.WriteFile(cfgPath, []byte(redisConfig), 0o600)
-	}()
+	require.NoError(t, loader.Start(t.Context(), os.Stdout))
+	require.NoError(t, os.WriteFile(cfgPath, []byte(redisConfig), 0o600))
 
-	<-ctx.Done()
+	require.Eventually(t, func() bool {
+		return reloads.Load() > 1
+	}, reloadTimeout, reloadTick)
+	t.Logf("config reloaded %d times", reloads.Load())
 }
 
 func TestConfigLoaderStandaloneExtensionServerAndCustomResource(t *testing.T) {
-	tmpDir, err := os.MkdirTemp("", "envoy-gateway-configloader-test")
-	require.NoError(t, err)
-	defer func(path string) {
-		_ = os.RemoveAll(path)
-	}(tmpDir)
-
+	tmpDir := t.TempDir()
 	cfgPath := tmpDir + "/config.yaml"
 	require.NoError(t, os.WriteFile(cfgPath, []byte(standaloneConfig), 0o600))
 	s, err := config.New(os.Stdout, os.Stderr)
 	require.NoError(t, err)
 
-	ctx, cancel := context.WithCancel(context.Background())
-	defer func() {
-		cancel()
-	}()
-
-	type testResult struct {
-		changed int32
-		extMgr  *egv1a1.ExtensionManager
-	}
-	resultChannel := make(chan testResult, 1)
-
-	var changed int32
-	loader := New(cfgPath, s, func(hookCtx context.Context, cfg *config.Server) error {
-		// Check if the hook context is canceled before incrementing
-		select {
-		case <-hookCtx.Done():
-			return nil
-		default:
-		}
-		c := atomic.AddInt32(&changed, 1)
-		t.Logf("config changed %d times", c)
-		if c > 1 {
-			resultChannel <- testResult{changed: c, extMgr: cfg.EnvoyGateway.ExtensionManager}
-			cancel()
-		}
+	var reloads atomic.Int32
+	loader := New(cfgPath, s, func(context.Context, *config.Server) error {
+		reloads.Add(1)
 		return nil
 	})
 
-	require.NoError(t, loader.Start(ctx, os.Stdout))
+	require.NoError(t, loader.Start(t.Context(), os.Stdout))
 	require.NotNil(t, loader.cfg.EnvoyGateway)
 	require.Nil(t, loader.cfg.EnvoyGateway.ExtensionManager)
 
-	go func() {
-		_ = os.WriteFile(cfgPath, []byte(standaloneConfigWithExtensionServer), 0o600)
-	}()
+	require.NoError(t, os.WriteFile(cfgPath, []byte(standaloneConfigWithExtensionServer), 0o600))
 
-	<-ctx.Done()
-	res := <-resultChannel // wait until second reload
-	require.Equal(t, int32(2), res.changed)
-	require.NotNil(t, res.extMgr)
-	require.NotNil(t, res.extMgr.PolicyResources)
-	require.Len(t, res.extMgr.PolicyResources, 1)
-	require.Equal(t, "gateway.example.io", res.extMgr.PolicyResources[0].Group)
-	require.Equal(t, "v1alpha1", res.extMgr.PolicyResources[0].Version)
-	require.Equal(t, "ExampleExtPolicy", res.extMgr.PolicyResources[0].Kind)
+	// Wait for the reload to apply the extension server config.
+	var extMgr *egv1a1.ExtensionManager
+	require.Eventually(t, func() bool {
+		cfg := loader.snapshotConfig()
+		if cfg == nil || cfg.EnvoyGateway == nil || cfg.EnvoyGateway.ExtensionManager == nil {
+			return false
+		}
+		extMgr = cfg.EnvoyGateway.ExtensionManager
+		return reloads.Load() > 1
+	}, reloadTimeout, reloadTick)
+	t.Logf("config reloaded %d times", reloads.Load())
+
+	require.Greater(t, reloads.Load(), int32(1))
+	require.NotNil(t, extMgr)
+	require.NotNil(t, extMgr.PolicyResources)
+	require.Len(t, extMgr.PolicyResources, 1)
+	require.Equal(t, "gateway.example.io", extMgr.PolicyResources[0].Group)
+	require.Equal(t, "v1alpha1", extMgr.PolicyResources[0].Version)
+	require.Equal(t, "ExampleExtPolicy", extMgr.PolicyResources[0].Kind)
 }
 
 // TestConfigLoaderDefaultsBeforeValidate ensures the hot-reload path applies defaults
@@ -137,53 +105,30 @@ func TestConfigLoaderStandaloneExtensionServerAndCustomResource(t *testing.T) {
 // Reloading to such a config must therefore succeed and surface the defaulted values
 // to the hook.
 func TestConfigLoaderDefaultsBeforeValidate(t *testing.T) {
-	tmpDir, err := os.MkdirTemp("", "envoy-gateway-configloader-test")
-	require.NoError(t, err)
-	defer func(path string) {
-		_ = os.RemoveAll(path)
-	}(tmpDir)
-
+	tmpDir := t.TempDir()
 	cfgPath := tmpDir + "/config.yaml"
 	require.NoError(t, os.WriteFile(cfgPath, []byte(defaultConfig), 0o600))
 	s, err := config.New(os.Stdout, os.Stderr)
 	require.NoError(t, err)
 
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	type testResult struct {
-		changed int32
-		eg      *egv1a1.EnvoyGateway
-	}
-	resultChannel := make(chan testResult, 1)
-
-	var changed int32
-	loader := New(cfgPath, s, func(hookCtx context.Context, cfg *config.Server) error {
-		select {
-		case <-hookCtx.Done():
-			return nil
-		default:
-		}
-		c := atomic.AddInt32(&changed, 1)
-		if c > 1 {
-			resultChannel <- testResult{changed: c, eg: cfg.EnvoyGateway}
-			cancel()
-		}
+	var reloads atomic.Int32
+	loader := New(cfgPath, s, func(context.Context, *config.Server) error {
+		reloads.Add(1)
 		return nil
 	})
 
-	require.NoError(t, loader.Start(ctx, os.Stdout))
-	go func() {
-		_ = os.WriteFile(cfgPath, []byte(minimalConfig), 0o600)
-	}()
+	require.NoError(t, loader.Start(t.Context(), os.Stdout))
+	require.NoError(t, os.WriteFile(cfgPath, []byte(minimalConfig), 0o600))
+	require.Eventually(t, func() bool {
+		return reloads.Load() > 1
+	}, reloadTimeout, reloadTick)
+	t.Logf("config reloaded %d times", reloads.Load())
 
-	<-ctx.Done()
-	res := <-resultChannel
-	require.Equal(t, int32(2), res.changed)
-	require.NotNil(t, res.eg)
-	require.NotNil(t, res.eg.Gateway)
-	require.Equal(t, egv1a1.GatewayControllerName, res.eg.Gateway.ControllerName)
-	require.NotNil(t, res.eg.Provider)
-	require.Equal(t, egv1a1.ProviderTypeKubernetes, res.eg.Provider.Type)
-	require.NotNil(t, res.eg.Logging)
+	eg := loader.snapshotConfig().EnvoyGateway
+	require.NotNil(t, eg)
+	require.NotNil(t, eg.Gateway)
+	require.Equal(t, egv1a1.GatewayControllerName, eg.Gateway.ControllerName)
+	require.NotNil(t, eg.Provider)
+	require.Equal(t, egv1a1.ProviderTypeKubernetes, eg.Provider.Type)
+	require.NotNil(t, eg.Logging)
 }

--- a/internal/envoygateway/config/loader/configloader_test.go
+++ b/internal/envoygateway/config/loader/configloader_test.go
@@ -11,6 +11,7 @@ import (
 	"os"
 	"sync/atomic"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/require"
 
@@ -31,159 +32,102 @@ var (
 	minimalConfig string
 )
 
-func TestConfigLoader(t *testing.T) {
-	tmpDir, err := os.MkdirTemp("", "envoy-gateway-configloader-test")
-	require.NoError(t, err)
-	defer func(path string) {
-		_ = os.RemoveAll(path)
-	}(tmpDir)
+const (
+	reloadTimeout = 30 * time.Second
+	reloadTick    = 10 * time.Millisecond
+)
 
+func TestConfigLoader(t *testing.T) {
+	tmpDir := t.TempDir()
 	cfgPath := tmpDir + "/config.yaml"
 	require.NoError(t, os.WriteFile(cfgPath, []byte(defaultConfig), 0o600))
 	s, err := config.New(os.Stdout, os.Stderr)
 	require.NoError(t, err)
 
-	ctx, cancel := context.WithCancel(context.Background())
-	defer func() {
-		cancel()
-	}()
-
-	changed := 0
-	loader := New(cfgPath, s, func(hookCtx context.Context, _ *config.Server) error {
-		changed++
-		// Only log if the hook context is still active to avoid panic
-		select {
-		case <-hookCtx.Done():
-			return nil
-		default:
-			t.Logf("config changed %d times", changed)
-		}
-		if changed > 1 {
-			cancel()
-		}
+	var reloads atomic.Int32
+	loader := New(cfgPath, s, func(context.Context, *config.Server) error {
+		reloads.Add(1)
 		return nil
 	})
 
-	require.NoError(t, loader.Start(ctx, os.Stdout))
-	go func() {
-		_ = os.WriteFile(cfgPath, []byte(redisConfig), 0o600)
-	}()
+	require.NoError(t, loader.Start(t.Context(), os.Stdout))
+	require.NoError(t, os.WriteFile(cfgPath, []byte(redisConfig), 0o600))
 
-	<-ctx.Done()
+	require.Eventually(t, func() bool {
+		return reloads.Load() > 1
+	}, reloadTimeout, reloadTick)
+	t.Logf("config reloaded %d times", reloads.Load())
 }
 
 func TestConfigLoaderStandaloneExtensionServerAndCustomResource(t *testing.T) {
-	tmpDir, err := os.MkdirTemp("", "envoy-gateway-configloader-test")
-	require.NoError(t, err)
-	defer func(path string) {
-		_ = os.RemoveAll(path)
-	}(tmpDir)
-
+	tmpDir := t.TempDir()
 	cfgPath := tmpDir + "/config.yaml"
 	require.NoError(t, os.WriteFile(cfgPath, []byte(standaloneConfig), 0o600))
 	s, err := config.New(os.Stdout, os.Stderr)
 	require.NoError(t, err)
 
-	ctx, cancel := context.WithCancel(context.Background())
-	defer func() {
-		cancel()
-	}()
-
-	type testResult struct {
-		changed int32
-		extMgr  *egv1a1.ExtensionManager
-	}
-	resultChannel := make(chan testResult, 1)
-
-	var changed int32
-	loader := New(cfgPath, s, func(hookCtx context.Context, cfg *config.Server) error {
-		// Check if the hook context is canceled before incrementing
-		select {
-		case <-hookCtx.Done():
-			return nil
-		default:
-		}
-		c := atomic.AddInt32(&changed, 1)
-		t.Logf("config changed %d times", c)
-		if c > 1 {
-			resultChannel <- testResult{changed: c, extMgr: cfg.EnvoyGateway.ExtensionManager}
-			cancel()
-		}
+	var reloads atomic.Int32
+	loader := New(cfgPath, s, func(context.Context, *config.Server) error {
+		reloads.Add(1)
 		return nil
 	})
 
-	require.NoError(t, loader.Start(ctx, os.Stdout))
+	require.NoError(t, loader.Start(t.Context(), os.Stdout))
 	require.NotNil(t, loader.cfg.EnvoyGateway)
 	require.Nil(t, loader.cfg.EnvoyGateway.ExtensionManager)
 
-	go func() {
-		_ = os.WriteFile(cfgPath, []byte(standaloneConfigWithExtensionServer), 0o600)
-	}()
+	require.NoError(t, os.WriteFile(cfgPath, []byte(standaloneConfigWithExtensionServer), 0o600))
 
-	<-ctx.Done()
-	res := <-resultChannel // wait until second reload
-	require.Equal(t, int32(2), res.changed)
-	require.NotNil(t, res.extMgr)
-	require.NotNil(t, res.extMgr.PolicyResources)
-	require.Len(t, res.extMgr.PolicyResources, 1)
-	require.Equal(t, "gateway.example.io", res.extMgr.PolicyResources[0].Group)
-	require.Equal(t, "v1alpha1", res.extMgr.PolicyResources[0].Version)
-	require.Equal(t, "ExampleExtPolicy", res.extMgr.PolicyResources[0].Kind)
+	// Wait for the reload to apply the extension server config.
+	var extMgr *egv1a1.ExtensionManager
+	require.Eventually(t, func() bool {
+		cfg := loader.snapshotConfig()
+		if cfg == nil || cfg.EnvoyGateway == nil || cfg.EnvoyGateway.ExtensionManager == nil {
+			return false
+		}
+		extMgr = cfg.EnvoyGateway.ExtensionManager
+		return reloads.Load() > 1
+	}, reloadTimeout, reloadTick)
+	t.Logf("config reloaded %d times", reloads.Load())
+
+	require.NotNil(t, extMgr)
+	require.NotNil(t, extMgr.PolicyResources)
+	require.Len(t, extMgr.PolicyResources, 1)
+	require.Equal(t, "gateway.example.io", extMgr.PolicyResources[0].Group)
+	require.Equal(t, "v1alpha1", extMgr.PolicyResources[0].Version)
+	require.Equal(t, "ExampleExtPolicy", extMgr.PolicyResources[0].Kind)
 }
 
 // TestConfigLoaderDefaultsBeforeValidate ensures the hot-reload path applies defaults
 // before validation. A config that omits `gateway` and `provider` fails validation
 // outright (`gateway is unspecified`) but is valid once defaults populate those fields.
 // Reloading to such a config must therefore succeed and surface the defaulted values
-// to the hook.
+// via the loader's config snapshot.
 func TestConfigLoaderDefaultsBeforeValidate(t *testing.T) {
-	tmpDir, err := os.MkdirTemp("", "envoy-gateway-configloader-test")
-	require.NoError(t, err)
-	defer func(path string) {
-		_ = os.RemoveAll(path)
-	}(tmpDir)
-
+	tmpDir := t.TempDir()
 	cfgPath := tmpDir + "/config.yaml"
 	require.NoError(t, os.WriteFile(cfgPath, []byte(defaultConfig), 0o600))
 	s, err := config.New(os.Stdout, os.Stderr)
 	require.NoError(t, err)
 
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	type testResult struct {
-		changed int32
-		eg      *egv1a1.EnvoyGateway
-	}
-	resultChannel := make(chan testResult, 1)
-
-	var changed int32
-	loader := New(cfgPath, s, func(hookCtx context.Context, cfg *config.Server) error {
-		select {
-		case <-hookCtx.Done():
-			return nil
-		default:
-		}
-		c := atomic.AddInt32(&changed, 1)
-		if c > 1 {
-			resultChannel <- testResult{changed: c, eg: cfg.EnvoyGateway}
-			cancel()
-		}
+	var reloads atomic.Int32
+	loader := New(cfgPath, s, func(context.Context, *config.Server) error {
+		reloads.Add(1)
 		return nil
 	})
 
-	require.NoError(t, loader.Start(ctx, os.Stdout))
-	go func() {
-		_ = os.WriteFile(cfgPath, []byte(minimalConfig), 0o600)
-	}()
+	require.NoError(t, loader.Start(t.Context(), os.Stdout))
+	require.NoError(t, os.WriteFile(cfgPath, []byte(minimalConfig), 0o600))
+	require.Eventually(t, func() bool {
+		return reloads.Load() > 1
+	}, reloadTimeout, reloadTick)
+	t.Logf("config reloaded %d times", reloads.Load())
 
-	<-ctx.Done()
-	res := <-resultChannel
-	require.Equal(t, int32(2), res.changed)
-	require.NotNil(t, res.eg)
-	require.NotNil(t, res.eg.Gateway)
-	require.Equal(t, egv1a1.GatewayControllerName, res.eg.Gateway.ControllerName)
-	require.NotNil(t, res.eg.Provider)
-	require.Equal(t, egv1a1.ProviderTypeKubernetes, res.eg.Provider.Type)
-	require.NotNil(t, res.eg.Logging)
+	eg := loader.snapshotConfig().EnvoyGateway
+	require.NotNil(t, eg)
+	require.NotNil(t, eg.Gateway)
+	require.Equal(t, egv1a1.GatewayControllerName, eg.Gateway.ControllerName)
+	require.NotNil(t, eg.Provider)
+	require.Equal(t, egv1a1.ProviderTypeKubernetes, eg.Provider.Type)
+	require.NotNil(t, eg.Logging)
 }

